File: cia-core/src/main/java/top/ourisland/creepersiarena/core/command/handler/player/PlayerPreferenceHandlers.java
package top.ourisland.creepersiarena.core.command.handler.player;

import org.bukkit.command.CommandSender;
import top.ourisland.creepersiarena.api.economy.cosmetic.CosmeticSlot;
import top.ourisland.creepersiarena.api.economy.cosmetic.ICosmeticService;
import top.ourisland.creepersiarena.core.bootstrap.BootstrapRuntime;
import top.ourisland.creepersiarena.core.command.handler.CommandHandlerContext;
import top.ourisland.creepersiarena.core.command.message.CommandMessenger;
import top.ourisland.creepersiarena.core.command.message.CommandPanel;
import top.ourisland.creepersiarena.core.command.message.CommandUsage;
import top.ourisland.creepersiarena.core.command.service.PlayerPreferenceService;
import top.ourisland.creepersiarena.core.command.service.UserLanguageService;

import static top.ourisland.creepersiarena.core.command.CommandParsers.asPlayer;
import static top.ourisland.creepersiarena.core.command.CommandParsers.parseBoolean;

public final class PlayerPreferenceHandlers {

    private final BootstrapRuntime rt;
    private final CommandMessenger messenger;

    public PlayerPreferenceHandlers(CommandHandlerContext context) {
        this.rt = context.runtime();
        this.messenger = context.messenger();
    }

    public void preferenceUsage(CommandSender sender) {
        messenger.panel(sender, CommandPanel.builder("Preferences")
                .row(new CommandUsage("/cia pref language <default|en_us|zh_cn>", "Change your personal language.").toMiniRow())
                .row(new CommandUsage("/cia pref particles <true|false>", "Enable or disable particle cosmetics for yourself.").toMiniRow())
                .row(new CommandUsage("/cia pref scoreboard <true|false>", "Enable or disable scoreboard UI preference.").toMiniRow())
                .row(new CommandUsage("/cia pref reset", "Reset command preferences to defaults.").toMiniRow())
                .build());
    }

    public void preferenceLanguage(CommandSender sender, String languageId) {
        language(sender, languageId);
    }

    public void language(CommandSender sender, String languageId) {
        var playerOpt = asPlayer(sender);
        if (playerOpt.isEmpty()) return;
        var p = playerOpt.get();

        var ul = rt.requireService(UserLanguageService.class);

        if (languageId.equalsIgnoreCase("default")) {
            ul.set(p, null);
            messenger.success(p, "Language reset to the server default.");
            return;
        }

        ul.set(p, languageId);
        messenger.successMini(p, "Language set to: " + messenger.id(languageId));
    }

    public void preferenceLanguageUsage(CommandSender sender) {
        languageUsage(sender);
    }

    public void languageUsage(CommandSender sender) {
        messenger.panel(sender, CommandPanel.builder("Language")
                .row(new CommandUsage("/cia language default", "Use the server default language.").toMiniRow())
                .row(new CommandUsage("/cia language en_us", "Use English.").toMiniRow())
                .row(new CommandUsage("/cia language zh_cn", "Use Simplified Chinese.").toMiniRow())
                .build());
    }

    public void preferenceParticles(CommandSender sender, String token) {
        var enabled = parseBoolean(token);
        if (enabled == null) {
            messenger.errorMini(sender, "Invalid preference value: " + messenger.id(token));
            messenger.hint(sender, "Use true/false or on/off.");
            return;
        }
        preferenceParticles(sender, enabled);
    }

    public void preferenceParticles(CommandSender sender, boolean enabled) {
        var playerOpt = asPlayer(sender);
        if (playerOpt.isEmpty()) return;
        var player = playerOpt.get();

        var preferences = rt.requireService(PlayerPreferenceService.class);
        preferences.particlesEnabled(player, enabled);
        if (!enabled) {
            var cosmetics = rt.getService(ICosmeticService.class);
            if (cosmetics != null && cosmetics.loaded(player.getUniqueId())) {
                cosmetics.clearSelection(player.getUniqueId(), CosmeticSlot.PARTICLE_TRAIL);
            }
        }
        messenger.successMini(sender, "Particle preference is now " + messenger.bool(enabled));
        if (!enabled) messenger.hint(sender, "Your current particle trail selection was cleared.");
    }

    public void preferenceScoreboard(CommandSender sender, String token) {
        var enabled = parseBoolean(token);
        if (enabled == null) {
            messenger.errorMini(sender, "Invalid preference value: " + messenger.id(token));
            messenger.hint(sender, "Use true/false or on/off.");
            return;
        }
        preferenceScoreboard(sender, enabled);
    }

    public void preferenceScoreboard(CommandSender sender, boolean enabled) {
        var playerOpt = asPlayer(sender);
        if (playerOpt.isEmpty()) return;
        var player = playerOpt.get();

        var preferences = rt.requireService(PlayerPreferenceService.class);
        preferences.scoreboardEnabled(player, enabled);
        messenger.successMini(sender, "Scoreboard preference is now " + messenger.bool(enabled));
        messenger.hint(sender, "This preference is saved and ready for scoreboard integrations.");
    }

    public void preferenceReset(CommandSender sender) {
        var playerOpt = asPlayer(sender);
        if (playerOpt.isEmpty()) return;
        var player = playerOpt.get();

        rt.requireService(PlayerPreferenceService.class).reset(player);
        rt.requireService(UserLanguageService.class).set(player, null);
        messenger.success(sender, "Preferences reset to defaults.");
        preference(sender);
    }

    public void preference(CommandSender sender) {
        var playerOpt = asPlayer(sender);
        if (playerOpt.isEmpty()) return;
        var player = playerOpt.get();

        var preferences = rt.requireService(PlayerPreferenceService.class);
        var languages = rt.requireService(UserLanguageService.class);
        var currentLanguage = languages.getOrNull(player);

        messenger.panel(sender, CommandPanel.builder("Preferences")
                .row("<gray>Language:</gray> " + (currentLanguage == null
                        ? "<gold>server default</gold>"
                        : messenger.id(currentLanguage)))
                .row("<gray>Particles:</gray> " + messenger.bool(preferences.particlesEnabled(player)))
                .row("<gray>Scoreboard:</gray> " + messenger.bool(preferences.scoreboardEnabled(player)))
                .row("")
                .row(new CommandUsage("/cia pref language <default|en_us|zh_cn>", "Change your personal language.").toMiniRow())
                .row(new CommandUsage("/cia pref particles <true|false>", "Enable or disable particle cosmetics for yourself.").toMiniRow())
                .row(new CommandUsage("/cia pref scoreboard <true|false>", "Enable or disable scoreboard UI preference.").toMiniRow())
                .row(new CommandUsage("/cia pref reset", "Reset command preferences to defaults.").toMiniRow())
                .build());
    }

}

File: cia-core/src/main/java/top/ourisland/creepersiarena/core/command/handler/player/PlayerPreferenceHandlers.kt
package top.ourisland.creepersiarena.core.command.handler.player

import org.bukkit.command.CommandSender
import top.ourisland.creepersiarena.api.economy.cosmetic.CosmeticSlot
import top.ourisland.creepersiarena.api.economy.cosmetic.ICosmeticService
import top.ourisland.creepersiarena.core.bootstrap.BootstrapRuntime
import top.ourisland.creepersiarena.core.command.CommandParsers.asPlayer
import top.ourisland.creepersiarena.core.command.CommandParsers.parseBoolean
import top.ourisland.creepersiarena.core.command.handler.CommandHandlerContext
import top.ourisland.creepersiarena.core.command.message.CommandMessenger
import top.ourisland.creepersiarena.core.command.message.CommandPanel
import top.ourisland.creepersiarena.core.command.message.CommandUsage
import top.ourisland.creepersiarena.core.command.service.PlayerPreferenceService
import top.ourisland.creepersiarena.core.command.service.UserLanguageService

class PlayerPreferenceHandlers(
    context: CommandHandlerContext,
) {

    private val rt: BootstrapRuntime = context.runtime
    private val messenger: CommandMessenger = context.messenger

    fun preferenceUsage(sender: CommandSender) {
        messenger.panel(
            sender,
            CommandPanel.builder("Preferences")
                .row(
                    CommandUsage(
                        "/cia pref language <default|en_us|zh_cn>",
                        "Change your personal language."
                    ).toMiniRow()
                )
                .row(
                    CommandUsage(
                        "/cia pref particles <true|false>",
                        "Enable or disable particle cosmetics for yourself."
                    ).toMiniRow()
                )
                .row(
                    CommandUsage(
                        "/cia pref scoreboard <true|false>",
                        "Enable or disable scoreboard UI preference."
                    ).toMiniRow()
                )
                .row(
                    CommandUsage(
                        "/cia pref reset", "Reset command preferences to defaults."
                    ).toMiniRow()
                )
                .build()
        )
    }

    fun preferenceLanguage(sender: CommandSender, languageId: String) {
        language(sender, languageId)
    }

    fun language(sender: CommandSender, languageId: String) {
        val playerOpt = asPlayer(sender)
        if (playerOpt.isEmpty) return
        val p = playerOpt.get()

        val ul = rt.requireService(UserLanguageService::class.java)

        if (languageId.equals("default", ignoreCase = true)) {
            ul.set(p, null)
            messenger.success(p, "Language reset to the server default.")
            return
        }

        ul.set(p, languageId)
        messenger.successMini(p, "Language set to: ${messenger.id(languageId)}")
    }

    fun preferenceLanguageUsage(sender: CommandSender) {
        languageUsage(sender)
    }

    fun languageUsage(sender: CommandSender) {
        messenger.panel(
            sender,
            CommandPanel.builder("Language")
                .row(
                    CommandUsage(
                        "/cia language default", "Use the server default language."
                    ).toMiniRow()
                )
                .row(
                    CommandUsage(
                        "/cia language en_us", "Use English."
                    ).toMiniRow()
                )
                .row(
                    CommandUsage(
                        "/cia language zh_cn", "Use Simplified Chinese."
                    ).toMiniRow()
                )
                .build()
        )
    }

    fun preferenceParticles(sender: CommandSender, token: String) {
        val enabled = parseBoolean(token)
        if (enabled == null) {
            messenger.errorMini(sender, "Invalid preference value: ${messenger.id(token)}")
            messenger.hint(sender, "Use true/false or on/off.")
            return
        }
        preferenceParticles(sender, enabled)
    }

    fun preferenceParticles(sender: CommandSender, enabled: Boolean) {
        val playerOpt = asPlayer(sender)
        if (playerOpt.isEmpty) return
        val player = playerOpt.get()

        val preferences = rt.requireService(PlayerPreferenceService::class.java)
        preferences.particlesEnabled(player, enabled)
        if (!enabled) {
            val cosmetics = rt.getService(ICosmeticService::class.java)
            if (cosmetics != null && cosmetics.loaded(player.uniqueId)) {
                cosmetics.clearSelection(player.uniqueId, CosmeticSlot.PARTICLE_TRAIL)
            }
        }
        messenger.successMini(sender, "Particle preference is now ${messenger.bool(enabled)}")
        if (!enabled) messenger.hint(sender, "Your current particle trail selection was cleared.")
    }

    fun preferenceScoreboard(sender: CommandSender, token: String) {
        val enabled = parseBoolean(token)
        if (enabled == null) {
            messenger.errorMini(sender, "Invalid preference value: ${messenger.id(token)}")
            messenger.hint(sender, "Use true/false or on/off.")
            return
        }
        preferenceScoreboard(sender, enabled)
    }

    fun preferenceScoreboard(sender: CommandSender, enabled: Boolean) {
        val playerOpt = asPlayer(sender)
        if (playerOpt.isEmpty) return
        val player = playerOpt.get()

        val preferences = rt.requireService(PlayerPreferenceService::class.java)
        preferences.scoreboardEnabled(player, enabled)
        messenger.successMini(sender, "Scoreboard preference is now ${messenger.bool(enabled)}")
        messenger.hint(sender, "This preference is saved and ready for scoreboard integrations.")
    }

    fun preferenceReset(sender: CommandSender) {
        val playerOpt = asPlayer(sender)
        if (playerOpt.isEmpty) return
        val player = playerOpt.get()

        rt.requireService(PlayerPreferenceService::class.java).reset(player)
        rt.requireService(UserLanguageService::class.java).set(player, null)
        messenger.success(sender, "Preferences reset to defaults.")
        preference(sender)
    }

    fun preference(sender: CommandSender) {
        val playerOpt = asPlayer(sender)
        if (playerOpt.isEmpty) return
        val player = playerOpt.get()

        val preferences = rt.requireService(PlayerPreferenceService::class.java)
        val languages = rt.requireService(UserLanguageService::class.java)
        val currentLanguage = languages.getOrNull(player)

        messenger.panel(
            sender,
            CommandPanel.builder("Preferences")
                .row("<gray>Language:</gray> " + (currentLanguage?.let { messenger.id(it) }
                    ?: "<gold>server default</gold>"))
                .row("<gray>Particles:</gray> ${messenger.bool(preferences.particlesEnabled(player))}")
                .row("<gray>Scoreboard:</gray> ${messenger.bool(preferences.scoreboardEnabled(player))}")
                .row("")
                .row(
                    CommandUsage(
                        "/cia pref language <default|en_us|zh_cn>",
                        "Change your personal language."
                    ).toMiniRow()
                )
                .row(
                    CommandUsage(
                        "/cia pref particles <true|false>",
                        "Enable or disable particle cosmetics for yourself."
                    ).toMiniRow()
                )
                .row(
                    CommandUsage(
                        "/cia pref scoreboard <true|false>",
                        "Enable or disable scoreboard UI preference."
                    ).toMiniRow()
                )
                .row(
                    CommandUsage(
                        "/cia pref reset", "Reset command preferences to defaults."
                    ).toMiniRow()
                )
                .build()
        )
    }

}
